User Model Cleanup & Refactoring

nanafox opened the pull request:
This pull request includes several changes to the Centa User Management Service project, focusing on enhancing the user and role management functionality, improving code readability, and updating documentation. The most important changes include the addition of new models and migrations, updates to the admin configuration, and various improvements to the `CustomUserManager` class.

### Enhancements to User and Role Management:

* **New Models and Migrations:**
  - Added models for `Organisation`, `Permission`, `Role`, `RolePermission`, `User`, and `UserProfile` in `src/user_service/models.py`.
  - Created initial migration file `0001_initial.py` to define the database schema for the new models.

* **Admin Configuration:**
  - Updated `src/user_service/admin.py` to register the new models with the Django admin site.

### Code Readability Improvements:

* **CustomUserManager Class:**
  - Improved the `create_user`, `create_superuser`, `deactivate_user`, `activate_user`, and `update_user` methods to use `ValidationError` instead of `ValueError` and added helper methods for better code organization. [[1]](diffhunk://#diff-4d15596f5b4bceb5073e471951cdbce6df82ae7dde571c74cbf393ff9b4c6b7cL25-R101) [[2]](diffhunk://#diff-4d15596f5b4bceb5073e471951cdbce6df82ae7dde571c74cbf393ff9b4c6b7cL94-R147)

* **General Code Cleanup:**
  - Removed unnecessary comments and improved docstrings for better clarity in `src/user_service/models.py`. [[1]](diffhunk://#diff-4d15596f5b4bceb5073e471951cdbce6df82ae7dde571c74cbf393ff9b4c6b7cL1-R10) [[2]](diffhunk://#diff-4d15596f5b4bceb5073e471951cdbce6df82ae7dde571c74cbf393ff9b4c6b7cL122-R292)

### Documentation Update:

* **README.md:**
  - Added instructions for creating a secret key and setting up the `.env` file for local development.

These changes collectively improve the functionality, maintainability, and documentation of the Centa User Management Service project.

nanafox committed:
chore: Update documentation with few styling

nanafox committed:
chore: Update requirements.txt file with update packages

nanafox committed:
feat: Register models in admin dashboard

nanafox committed:
refactor: Update models to fix errors and handle minor validations

nanafox committed:
feat: Add initial database model migration

thenoblet commented:
If a user lookup fails a UserDoesNotExist exception is to be raised. Validation errors are mostly for form validations, model field validations and custom validation logic.

thenoblet commented:
Also in the `update_user` method, I think we should check if the user model has an attribute before we set a value on the attribute. And, raise a ValueError if the field does not exist on the model.

What do you think??

thenoblet commented on src/user_service/models.py:
If a user lookup fails a UserDoesNotExist exception is to be raised. Validation errors are mostly for form validations, model field validations and for custom validation logic

nanafox replied:
Before this update, your initial code was raising a `ValueError`. I will update it to raise the appropriate exception now.

thenoblet reviewed (approved):
Looks good... well done, Mentor!!!

nanafox commented:
> Also in the `update_user` method, I think we should check if the user model has an attribute before we set a value on the attribute. And, raise a ValueError if the field does not exist on the model.
> 
> What do you think??

We could do that but instead of `ValueError`, it will be an `AttributeError` since we are setting them as attributes of the that user instance.

If this is the case, we will also need to have the same checks in the `create_user` method as well since it also accepts multiple extra fields.

nanafox commented:
> > Also in the `update_user` method, I think we should check if the user model has an attribute before we set a value on the attribute. And, raise a ValueError if the field does not exist on the model.
> > What do you think??
> 
> We could do that but instead of `ValueError`, it will be an `AttributeError` since we are setting them as attributes of the that user instance.
> 
> If this is the case, we will also need to have the same checks in the `create_user` method as well since it also accepts multiple extra fields.

Here's an example of how it would go. This example uses the `create_user` method, but it will work with the `update_user` method as well. This way, errors are silent when an invalid attribute is passed to either of these methods.

![image](https://github.com/user-attachments/assets/a46c6283-a167-4b98-8074-ffbf01bc92c9)

nanafox committed:
refactor: Raise `DoesNotExist` error for non-existing emails

This update raises `User.DoesNotExist` for user emails that does not
exist instead of `ValidationError`. This ensures best practices and
consistency.

Also, this update adds a check to verify that all atributes are valid
and exists while creating or updating user objects. The implementation
is in the `set_extra_attributes` method.

nanafox committed:
chore: Fix typo when throwing `AttributeError` in `set_extra_attributes`

thenoblet commented:
> > > Also in the `update_user` method, I think we should check if the user model has an attribute before we set a value on the attribute. And, raise a ValueError if the field does not exist on the model.
> > > What do you think??
> > 
> > 
> > We could do that but instead of `ValueError`, it will be an `AttributeError` since we are setting them as attributes of the that user instance.
> > If this is the case, we will also need to have the same checks in the `create_user` method as well since it also accepts multiple extra fields.
> 
> Here's an example of how it would go. This example uses the `create_user` method, but it will work with the `update_user` method as well. This way, errors are silent when an invalid attribute is passed to either of these methods.
> 
> ![image](https://private-user-images.githubusercontent.com/48143641/392737280-a46c6283-a167-4b98-8074-ffbf01bc92c9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzMzOTE0MTYsIm5iZiI6MTczMzM5MTExNiwicGF0aCI6Ii80ODE0MzY0MS8zOTI3MzcyODAtYTQ2YzYyODMtYTE2Ny00Yjk4LTgwNzQtZmZiZjAxYmM5MmM5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjA1VDA5MzE1NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA0ODU3NTNjNWIzMWQ2MjM1MjgxYWMyMGVhYzI3N2M5ZGRhM2Y4MmQ2OWY4M2M2MWNlMjk4N2EwY2E1M2Q0YzgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.LYPnrScqfcJUw9agy40bGNCYv9Z_rDFkYYoR3I2r9Jc)

Looks good!
